### PR TITLE
Added move_base plugins for costmaps

### PR DESCRIPTION
**Pull request by schmerl**
Added a plugin mechanism for move_base that loads in plugins for costmap layers that deal with the map, obstacles, and other aspects of the costmaps.

### CODE REVIEW

File: src/rosdiscover/models/image_transport.py [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+
+    for a in args:
+        if a in ["raw", "compressed", "theora"]:
+            img_format = a
+        elif a.startswith("in:="):
+            intopic = a[4:]
+        elif a.startswith("out:="):
+            outopic = a[5:]
+        else:
+            logger.error(f'"{a}" is not a valid argument for image_transport/republishZZZZ')
```
Comment (ChrisTimperley) [inline src/rosdiscover/models/image_transport.py:20]: Typo in this string?

File: src/rosdiscover/models/image_transport.py [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+
+@model('image_transport', 'republish')
+def republish(c):
+    args = c.args.split(' ')
+    intopic: str = 'in'
+    outopic: str = 'out'
+    img_format: str = 'raw'
+
+    for a in args:
+        if a in ["raw", "compressed", "theora"]:
```
Comment (ChrisTimperley) [inline src/rosdiscover/models/image_transport.py:13]: Super minor comment: You can change the list, which is mutable, to a tuple, which is immutable. In _theory_, this should be faster.

File: src/rosdiscover/models/move_base.py [hunk 3]
```diff
@@ -1,4 +1,7 @@
 from ..interpreter import model
+from loguru import logger
```
Comment (ChrisTimperley) [inline src/rosdiscover/models/move_base.py:2]: Apologies for the nitpicking comment: The general order of imports should be Python standard library stuff in one block at the top, followed by external imports in another block (e.g., `loguru`, and then, finally, relative imports in a last block: https://www.python.org/dev/peps/pep-0008/#id23

The linter doesn't seem to be clever enough to figure this out.

File: src/rosdiscover/models/move_base.py [hunk 4]
```diff
@@ -74,7 +78,7 @@ def base_plugin_local(name):
         # type: (str) -> None
         c.pub("~{}/global_plan".format(name), "nav_msgs/Path")
```
Comment (ChrisTimperley) [inline src/rosdiscover/models/move_base.py:79]: Prefer `f"{name}/global_plan"

File: src/rosdiscover/models/spawn_model.py [hunk 5]
```diff
@@ -39,10 +40,13 @@ def spawn_model(c):
     urdf_contents = c.read(urdf_param_name).strip()
     logger.debug(f'parsing URDF model from parameter [{urdf_param_name}]:'
                  f'\n{urdf_contents}')
-
     with open('urdf.xml', 'w') as f:
         f.write(urdf_contents)
-
+    # Non XML stuff can appear at the end, so strip it out by finding the position of the last
+    # end tag and then stripping everything after that tag
+    end_tags_endposition = [i.end() for i in re.finditer(r'</.*>', urdf_contents)]
```
Comment (ChrisTimperley) [inline src/rosdiscover/models/spawn_model.py:47]: `end_tags_end_position`

File: src/rosdiscover/models/topic_tools_mux.py [hunk 6]
```diff
@@ -21,3 +21,22 @@ def mux(c):
...
+
+@model("topic_tools", "relay")
+def relay(c):
+    c.read("~unreliable", False)
+    c.read('~lazy', False)
+    c.read('~stealth', False)
+    # TODO: Chris: how should we treat this? I think this means a topic is created
+    # that by default is the same is intopic, but could be set by a parameter
+    # http://wiki.ros.org/topic_tools/relay
+    c.read('~monitor_topic', 'intopic')
```
Comment (ChrisTimperley) [inline src/rosdiscover/models/topic_tools_mux.py:34]: `topic_name = c.read(....)`

File: src/rosdiscover/models/image_transport.py (hunk 1 above)
Comment (schmerl) [reply src/rosdiscover/models/image_transport.py]: Probably trying to exit vi.

File: src/rosdiscover/models/image_transport.py (hunk 2 above)
Comment (schmerl) [reply src/rosdiscover/models/image_transport.py]: Done.

File: src/rosdiscover/models/move_base.py (hunk 3 above)
Comment (schmerl) [reply src/rosdiscover/models/move_base.py:1]: Think done.

File: src/rosdiscover/models/move_base.py (hunk 4 above)
Comment (schmerl) [reply src/rosdiscover/models/move_base.py]: Fixed in all that code (not sure I wrote this originally, but it was a simple regexp replace)

File: src/rosdiscover/models/spawn_model.py (hunk 5 above)
Comment (schmerl) [reply src/rosdiscover/models/spawn_model.py]: done

File: src/rosdiscover/models/topic_tools_mux.py (hunk 6 above)
Comment (schmerl) [reply src/rosdiscover/models/topic_tools_mux.py]: Done. I am not sure of the weird semantics for this: http://wiki.ros.org/topic_tools/relay. But it does something now.

File: src/rosdiscover/models/image_transport.py [hunk 7]
```diff
@@ -0,0 +1,23 @@
+from ..interpreter import model
+from loguru import logger
```
Comment (ChrisTimperley) [inline src/rosdiscover/models/image_transport.py:2]: Same deal with loguru here.

File: src/rosdiscover/models/spawn_model.py [hunk 8]
```diff
@@ -6,6 +6,7 @@
 
 from .plugins.gazebo import GazeboPlugin
 from ..interpreter import model
+import re
```
Comment (ChrisTimperley) [inline src/rosdiscover/models/spawn_model.py:9]: Same thing with imports, sorry. Standard library imports go at the top.

Comment (schmerl) [reply src/rosdiscover/models/spawn_model.py]: Done

File: src/rosdiscover/models/image_transport.py (hunk 7 above)
Comment (schmerl) [reply src/rosdiscover/models/image_transport.py:1]: Done

**Comment by ChrisTimperley** [comment]
It looks like there's a few style issues in `topic_tools_mux`:
```
src/rosdiscover/models/topic_tools_mux.py:30:5: F841 local variable 'stealth' is assigned to but never used
src/rosdiscover/models/topic_tools_mux.py:32:5: F841 local variable 'topic' is assigned to but never used
src/rosdiscover/models/topic_tools_mux.py:41:5: E303 too many blank lines (2)
src/rosdiscover/models/topic_tools_mux.py:49:30: W292 no newline at end of file
src/rosdiscover/models/topic_tools_mux.py:49:30: W292 no newline at end of file
src/rosdiscover/models/topic_tools_mux.py:49:30: W292 no newline at end of file
```

**Comment by ChrisTimperley** [comment]
Looks like there's a few last style issues:
```
src/rosdiscover/models/topic_tools_mux.py:39:5: E303 too many blank lines (2)
src/rosdiscover/models/spawn_model.py:3:1: F401 're' imported but unused
```

**Comment by schmerl** [comment]
Should be fixed now.

**Pull request merged**